calyx/src/passes/infer_share.rs

```diff
@@ -0,0 +1,386 @@
+                {
+                    self.reads.insert(comp.clone_name());
+                }
+            }
+            ir::Control::If(ir::If { port, cond, .. }) => {
+                if is_begin_id {
+                    self.add_parent_if_shareable(state_shareable, port);
+                    if let Some(group) = cond {
+                        self.add_assignment_reads(
+                            state_shareable,
+                            &group.borrow().assignments,
+                        );
+                    }
+                }
+            }
+        }
+    }
+}
+
+///Contains the name of a cell. The methods in this struct are used to search to
+///see if there was a write to the cell name.
+struct NameSearch {
+    name: ir::Id,
+}
+
+impl NameSearch {
+    fn new(name: ir::Id) -> Self {
+        NameSearch { name }
+    }
+
+    //returns true if the port's parent's name matches name, false otherwise
+    fn parent_matches_name(&self, port: &ir::RRC<ir::Port>) -> bool {
+        if let ir::PortParent::Cell(cell) = &port.borrow().parent {
+            if cell.upgrade().borrow().name().clone() == self.name {
+                return true;
+            }
+        }
+        false
+    }
+
+    //returns Some(cell) if the assignment is to cell's go port and is guaranteed
+    //to be activated. To check this guarantee, we check if the assignment guard
+    //is true and assignment src is 1'd1.
+    fn guaranteed_go(assign: &ir::Assignment) -> Option<ir::RRC<ir::Cell>> {
+        let dst_ref = assign.dst.borrow();
+        if dst_ref.attributes.has("go")
+            && Self::guard_true(&assign.guard)
+            && Self::src_const(&assign.src)
+        {
+            if let ir::PortParent::Cell(cell_wref) = &dst_ref.parent {
+                return Some(Rc::clone(&cell_wref.upgrade()));
+            }
+        }
+        None
+    }
+
+    //returns true if port is 1'd1
+    fn src_const(port: &ir::RRC<ir::Port>) -> bool {
+        port.borrow().is_constant(1, 1)
+    }
+
+    //returns true if guard is True
+    fn guard_true(guard: &ir::Guard) -> bool {
+        matches!(guard, ir::Guard::True)
+    }
+
+    //given a vec of assignments, return true if name is ever written to in
+    //assignments, false otherwise
+    fn search_assignments(&self, assignments: &[ir::Assignment]) -> bool {
+        for write in assignments.iter().filter_map(Self::guaranteed_go) {
+            if *write.borrow().name() == self.name {
+                return true;
+            }
+        }
+        false
+    }
+
+    //Returns true if any of the control statements in dominators write to a cell
+    //with self's name.
+    fn is_written(
+        &self,
+        dominators: &HashSet<u64>,
+        comp: &mut ir::Component,
+        shareable: &ShareSet,
+    ) -> bool {
+        for dominator in dominators {
+            if let Some(c) =
+                DominatorMap::get_control(*dominator, &comp.control.borrow())
```

Comment:
It seems like a common operation on DominatorMap is to get a bunch of dominators and get the associated control nodes. Instead of iterating on the control tree over and over again, can we define a `DominatorMap::get_control_nodes` method that gets all of them at once? This make it easier for the API to get a bunch of control nodes in one go.